pkg/services/ngalert/notifier/channels/discord.go

```diff
@@ -54,78 +51,75 @@ type discordAttachment struct {
 
 const DiscordMaxEmbeds = 10
 
-func NewDiscordConfig(config *NotificationChannelConfig) (*DiscordConfig, error) {
-	discordURL := config.Settings.Get("url").MustString()
-	if discordURL == "" {
-		return nil, errors.New("could not find webhook url property in settings")
-	}
-	return &DiscordConfig{
-		NotificationChannelConfig: config,
-		Content:                   config.Settings.Get("message").MustString(DefaultMessageEmbed),
-		AvatarURL:                 config.Settings.Get("avatar_url").MustString(),
-		WebhookURL:                discordURL,
-		UseDiscordUsername:        config.Settings.Get("use_discord_username").MustBool(false),
-	}, nil
-}
-
 func DiscordFactory(fc FactoryConfig) (NotificationChannel, error) {
-	cfg, err := NewDiscordConfig(fc.Config)
+	dn, err := newDiscordNotifier(fc)
 	if err != nil {
 		return nil, receiverInitError{
 			Reason: err.Error(),
 			Cfg:    *fc.Config,
 		}
 	}
-	return NewDiscordNotifier(cfg, fc.NotificationService, fc.ImageStore, fc.Template), nil
+	return dn, nil
 }
 
-func NewDiscordNotifier(config *DiscordConfig, ns notifications.WebhookSender, images ImageStore, t *template.Template) *DiscordNotifier {
+func newDiscordNotifier(fc FactoryConfig) (*DiscordNotifier, error) {
+	settings := discordSettings{}
+	err := fc.Config.unmarshalSettings(&settings)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal settings: %w", err)
+	}
```

Comment:
Unfortunately I think the `bool` field in the settings can be hit by a similar issue as the other PR - see https://github.com/grafana/grafana/pull/57530#discussion_r1003646868

Although it's not quite as bad here, as in Discord there isn't a precedent with users about providing everything wrapped in strings like there is in the Pushover notifier.

Let's have a chat with @yuri-tceretian involved on how we can solve this problem. There's a second layer here regarding users that already have a notifier configured in their database. Since with `simplejson` incorrect types are ignored and not rejected, if we move to a more strongly-typed way of decoding, it can cause some existing configurations to fail. I suspect we can solve this with a database migration or by loosening the strictness of `encoding/json`.

Reply:
I did not see that coming. Thanks @alexweav!

Reply:
I thought of that but since, in tests, we are parsing incomplete json payloads as strings, we use `omitempty`, and we had no errors, it was safe to use.